sdks/java/extensions/euphoria/euphoria-beam/src/main/java/org/apache/beam/sdk/extensions/euphoria/beam/BeamExecutorContext.java

```diff
@@ -123,47 +131,112 @@ boolean strongTypingEnabled() {
     return false;
   }
 
+  /**
+   * Get return type of lambda function (e.g UnaryFunction).
+   *
+   * @return return type
+   */
+  static Type getLambdaReturnType(final Object lambda) {
+
+    Type returnType = null;
+    try {
+      final Method method = lambda.getClass().getDeclaredMethod("writeReplace");
+      method.setAccessible(true);
+      final SerializedLambda serializedLambda = SerializedLambda.class.cast(method.invoke(lambda));
+      final String signature = serializedLambda.getImplMethodSignature();
+      final MethodType methodType =
+          MethodType.fromMethodDescriptorString(signature, lambda.getClass().getClassLoader());
+      returnType = methodType.returnType();
+    } catch (IllegalAccessException
+        | IllegalArgumentException
+        | NoSuchMethodException
+        | InvocationTargetException e) {
+      // not a lambda try pure reflection, algo is trivial since it is functional interfaces: take
+      // the only not Object methods ;)
+      for (final Method m : lambda.getClass().getMethods()) {
```

Comment:
Functional interfaces may have default methods. Will this still work?

Reply:
I think it won't but it will fail to assign coder so it will call `getFallbackCoder`